File: 3336-water-bottles-ii/water-bottles-ii.java
class Solution {
    public int maxBottlesDrunk(int numBottles, int numExchange) 
    {
        int totalBottlesDrank=numBottles;
        int emptyBottles=numBottles;
        int fullBottles=0;
        while(fullBottles>0 || emptyBottles>=numExchange)
        {
            if(fullBottles>0)
            {
                totalBottlesDrank+= fullBottles;
                emptyBottles+=fullBottles;
                fullBottles--;
            }
            else
            {
                emptyBottles-=numExchange;
                fullBottles++;
                numExchange++;
            }
        }
        return totalBottlesDrank;
    }
}

File: 3336-water-bottles-ii/water-bottles-ii.cpp
class Solution {
public:
    int maxBottlesDrunk(int numBottles, int numExchange) 
    {
        int ans=numBottles;
        int empty=numBottles;
        numBottles=0;
        while(numBottles!=0 or empty>=numExchange)
        {
            if(empty>=numExchange)
            {
                numBottles++;
                empty-=numExchange;
                numExchange++;
            }
            else
            {
                empty+=numBottles;
                ans+=numBottles;
                numBottles=0;
            }
        }
        return ans;
    }
};